src/core/openthread.cpp

```diff
@@ -827,20 +880,34 @@ void HandleMleDiscover(otActiveScanResult *aResult, void *aContext)
     handler(aResult);
 }
 
-void otSetReceiveIp6DatagramCallback(otReceiveIp6DatagramCallback aCallback)
+void otSetReceiveIp6DatagramCallback(otContext *aContext, otReceiveIp6DatagramCallback aCallback, void *aCallbackContext)
 {
-    Ip6::Ip6::SetReceiveDatagramCallback(aCallback);
+    Ip6::Ip6::SetReceiveDatagramCallback(aContext, aCallback, aCallbackContext);
+}
+
+ThreadError otSendIp6Datagram(otContext *aContext, otMessage aMessage)
+{
+    otLogFuncEntry();
+    ThreadError error = 
+        Ip6::Ip6::HandleDatagram(
+            *static_cast<Message *>(aMessage), 
+            NULL, 
+            (uint8_t)aContext->mThreadNetif.GetInterfaceId(),
+            NULL, 
+            true
+            );
+    otLogFuncExitErr(error);
+    return error;
 }
 
-ThreadError otSendIp6Datagram(otMessage aMessage)
+otMessage otNewIPv6Message(otContext *aContext, uint16_t aLength)
 {
-    return Ip6::Ip6::HandleDatagram(*static_cast<Message *>(aMessage), NULL, sThreadNetif->GetInterfaceId(),
-                                    NULL, true);
+    return Ip6::Ip6::NewPrepopulatedMessage(aContext, aLength);
```

Comment:
Should we call SetLength on the message before we return it? Right now, the caller has to immediately call it, and it doesn't seem to make sense that if they pass in the length here, they should have to call another function to actually allocate the buffers. Is there a better way to preallocate all the necessary buffers?